tests/results_benchmark.py

```diff
@@ -137,62 +171,164 @@ def __init__(self, root: Path = Path("data/raw/pipal")):
 }
 
 
-def eval_metric(loader: torch.utils.data.DataLoader, metric: Callable, device: str) -> List:
+def eval_metric(loader: DataLoader, metric: Metric, device: str, feature_extractor: str) \
+        -> Tuple[np.ndarray, np.ndarray]:
     """Evaluate metric on a given dataset.
+
     Args:
         loader: PyTorch dataloader that returns batch of distorted images, reference images and scores.
-        metric: Should support `metric(x, y)` or `metric(x)` call.
+        metric: General metric that satisfies the Metric interface.
         device: Computation device.
-
+        feature_extractor: name of the neural network to be used to extract features from images
     Returns:
-        gt_scores: Ground truth values
+        gt_scores: Ground truth values.
         metric_scores: Predicted values as torch.Tensors.
     """
-    assert isinstance(loader, torch.utils.data.DataLoader), "Expect loader to be DataLoader class"
-    assert callable(metric), f"Expected metric to be callable, got {type(metric)} instead!"
+    assert isinstance(loader, DataLoader), "Expect loader to be DataLoader class"
+    assert isinstance(metric, Metric), f"Expected metric to be an instance of Metric, got {type(metric)} instead!"
 
     gt_scores = []
     metric_scores = []
+    compute_function = determine_compute_function(metric_category=metric.category)
 
-    for (distorted_images, reference_images, scores) in tqdm.tqdm(loader, ncols=50):
+    for distorted_images, reference_images, scores in tqdm.tqdm(loader, ncols=50):
         distorted_images, reference_images = distorted_images.to(device), reference_images.to(device)
         gt_scores.append(scores.cpu())
 
-        # Full Reference methods
-        metric_score = metric(distorted_images, reference_images).cpu()
+        metric_score: torch.Tensor = \
+            compute_function(metric.functor, distorted_images, reference_images, device, feature_extractor)
+
+        if np.isnan(metric_score):
+            metric_score = torch.Tensor([0])
```

Comment:
If metric returns NaN shouldn't we raise an exception? 
What a valid cases of NaN result that can be replaced with 0?
Why 0, some metrics should be minimised, some other - maximised, this can create confusion.

Reply:
I am observing problems with NaN values for some DB metrics (e.g., IS) on KADID10k dataset. I am not sure about the reason of such an output at this point in time. I do not consider this case to be valid.

Reply:
Removing this line (especially considering that it does not work with `batch_size > 1`). I think that we need to investigate the reasons of this behaviour in a different PR

Reply:
Check out the issue https://github.com/photosynthesis-team/piq/issues/309